### PR TITLE
Adding support for fetching resources based on requested resource types

### PR DESCRIPTION
**Pull request by prashanthpx**
Signed-off-by: Prashanth Kumar <prkumar@purestorage.com>

**What type of PR is this?**
Enhancement.

**What this PR does / why we need it**:
 Enhancement to support backing up of only particular resource types.

**Does this PR change a user-facing CRD or CLI?**:
Yes
**Is a release note needed?**:
None

**Does this change need to be cherry-picked to a release branch?**:
Yes for 2.6.3

### CODE REVIEW

File: pkg/resourcecollector/resourcecollector.go [hunk 1]
```diff
@@ -136,6 +143,134 @@ func resourceToBeCollected(resource metav1.APIResource, grp schema.GroupVersion,
...
+			objects.Items = append(objects.Items, runtimeObject)
+			objects.resourceMap[metadata.GetUID()] = true
+		}
+	}
+
+	modObjects, err := r.pruneOwnedResources(objects.Items, objects.resourceMap)
+	if err != nil {
+		return nil, err
+	}
+	objects.Items = modObjects
```
Comment (siva-portworx) [inline pkg/resourcecollector/resourcecollector.go:270]: I could not see prepareResourcesForCollection getting called in the GetResourcesForType(). Any reason? How will the individual resourceType prepare function will be called in this case?

File: pkg/resourcecollector/resourcecollector.go [hunk 2]
```diff
@@ -136,6 +143,134 @@ func resourceToBeCollected(resource metav1.APIResource, grp schema.GroupVersion,
...
+		}
+
+		var selectors string
+		// PVs don't get the labels from their PVCs, so don't use the label selector
+		// Also skip for some other resources that aren't necessarily tied to an application
+		switch resource.Kind {
+		case "PersistentVolume",
+			"ClusterRoleBinding",
+			"ClusterRole",
+			"ServiceAccount":
```
Comment (siva-portworx) [inline pkg/resourcecollector/resourcecollector.go:230]: Currently we are not skipping the ClusterRoleBinding, ClusterRole and service Account in GetResources. Probably, we can add this fix there as well.

File: pkg/applicationmanager/controllers/applicationbackup.go [hunk 3]
```diff
@@ -989,34 +1000,69 @@ func (a *ApplicationBackupController) backupResources(
...
+				for _, resource := range resourceTypes {
+					if resource.Kind == backupResourceType {
+						objects, err := a.resourceCollector.GetResourcesForType(resource, nil, batch, backup.Spec.Selectors, objectMap, true)
+						if err != nil {
+							log.ApplicationBackupLog(backup).Errorf("Error getting resources: %v", err)
+							break
+						}
+						allObjects = append(allObjects, objects.Items...)
+						// Do a dummy update to the backup CR to update only the last update timestamp
+						namespacedName := types.NamespacedName{}
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply pkg/applicationmanager/controllers/applicationbackup.go:1046]: Done.

File: pkg/resourcecollector/resourcecollector.go (hunk 1 above)
Comment (prashanthpx) [reply pkg/resourcecollector/resourcecollector.go:270]: Added.

File: pkg/resourcecollector/resourcecollector.go (hunk 2 above)
Comment (prashanthpx) [reply pkg/resourcecollector/resourcecollector.go]: Added

File: pkg/applicationmanager/controllers/applicationbackup.go [hunk 4]
```diff
@@ -989,17 +1000,33 @@ func (a *ApplicationBackupController) backupResources(
...
-			batch,
-			backup.Spec.Selectors,
-			objectMap,
-			optionalBackupResources,
-			true)
-		if err != nil {
-			log.ApplicationBackupLog(backup).Errorf("Error getting resources: %v", err)
-			return err
+		if len(backup.Spec.ResourceTypes) == 0 {
+			batch := namespacelist[i:min(i+backupResourcesBatchCount, len(namespacelist))]
```
Comment (siva-portworx) [inline pkg/applicationmanager/controllers/applicationbackup.go:1004]: This batch can be removed as we have one outside the if condition

Comment (prashanthpx) [reply pkg/applicationmanager/controllers/applicationbackup.go]: Yeah, overlooked it. Fixed it.

File: pkg/applicationmanager/controllers/applicationbackup.go [hunk 5]
```diff
@@ -979,6 +979,17 @@ func (a *ApplicationBackupController) backupResources(
...
+		resourceTypes, err = a.resourceCollector.GetResourceTypes(optionalResourceTypes, true)
+		if err != nil {
+			log.ApplicationBackupLog(backup).Errorf("Error getting resource types: %v", err)
+			return err
+		}
+	}
+
 	// Always backup optional resources. When restorting they need to be
 	// explicitly added to the spec
 	objectMap := stork_api.CreateObjectsMap(backup.Spec.IncludeResources)
```
Comment (siva-portworx) [inline pkg/applicationmanager/controllers/applicationbackup.go:995]: I think, backup.Spec.ResourceType list and backup.Spec.IncludeResources list are mutual exclusive. We might need to add a check such that both are not set.
Also, While we should pass includeResource list to GetResourcesForType() as it is based on resource type.

**Review by siva-portworx — APPROVED** [review]

**Pull request merged**